File: find_lane_lines.py
import numpy as np
import cv2
import math
from math import pi
from accumLines import *
from scipy.linalg import lstsq
from scipy.optimize import least_squares


class History:
    # This class will store the results from the past several image frames
    # and compute the average. Values stored are the voting landscape and
    # the end points for the left and right lane lines.

    def __init__(self, history_length, equal_weight=True):
        # Initialize storage for the current state of the voting landscape,
        # the left lane, and the right lane
        self.VL = np.zeros(0)  # np.zeros( img_size )
        self.left_lane = np.zeros((2, 2))  # [p_start; p_end]
        self.right_lane = np.zeros((2, 2))  # [p_start; p_end]

        # Initialize storage for the hisotry of values
        self.VL_hist = np.zeros(0)  # np.zeros( img_size + (history_length,) )
        self.left_lane_hist = np.zeros((2, 2) + (history_length,))
        self.right_lane_hist = np.zeros((2, 2) + (history_length,))

        if equal_weight:
            w = np.ones(history_length)
        else:
            w = np.linspace(0.3, 1, history_length)
        self.__weight = w  # .reshape((1,1,history_length))

        self.__history_length = history_length
        self._cntr = 0

        self.mask = []

    def add(self, VL, left_lane, right_lane):
        # Now that we know the size of the image we can initialize the
        # variables for the voting landscape
        if self._cntr == 0:
            img_size = VL.shape
            self.VL = np.zeros(img_size)
            self.VL_hist = np.zeros(img_size + (self.__history_length,))

        # Add to history
        self.VL_hist[:, :, 0] = VL
        self.VL_hist = np.roll(self.VL_hist, -1, axis=2)

        self.left_lane_hist[:, :, 0] = left_lane
        self.left_lane_hist = np.roll(self.left_lane_hist, -1, axis=2)

        self.right_lane_hist[:, :, 0] = right_lane
        self.right_lane_hist = np.roll(self.right_lane_hist, -1, axis=2)

        self._cntr += 1

        # Copmute new current values as mean of history
        if self._cntr < self.__history_length:
            W = np.sum(self.__weight[-self._cntr:])
            self.VL = self.VL_hist[
                :, :, -self._cntr:].dot(self.__weight[-self._cntr:]) / W
            self.left_lane = self.left_lane_hist[
                :, :, -self._cntr:].dot(self.__weight[-self._cntr:]) / W
            self.right_lane = self.right_lane_hist[
                :, :, -self._cntr:].dot(self.__weight[-self._cntr:]) / W
        else:
            W = np.sum(self.__weight)
            self.VL = self.VL_hist.dot(self.__weight) / W
            self.left_lane = self.left_lane_hist.dot(self.__weight) / W
            self.right_lane = self.right_lane_hist.dot(self.__weight) / W


def pre_process_image(I, roi, resize_factor=1):
    # Pre-process the images :
    # - resize them
    # - adjust contrast in ROI along rows
    # - smooth the image
    I = I.astype(np.float32)

    I = np.mean(I[:, :, :2], axis=2)

    # Resize image and ROI
    if resize_factor != 1:
        I = cv2.resize(I, (0, 0), fx=resize_factor, fy=resize_factor)
        roi = cv2.resize(roi.astype(np.float32), (0, 0),
                         fx=resize_factor, fy=resize_factor)
        roi = roi > 0

    # Adjust image contrast in ROI along rows
    tmpI = I.copy()
    # if np.issubsctype(I.dtype, np.integer):
    #     tmpI[np.logical_not(roi)] = np.iinfo(I.dtype).max
    # else:
    #     tmpI[np.logical_not(roi)] = np.Inf

    tmpI[np.logical_not(roi)] = np.Inf
    minI = np.percentile(tmpI, 30, axis=1, keepdims=True)
    I = I - minI  # Note, should probably smooth out minI with a moving average before subtracting
    I[I < 0] = 0

    tmpI = I.copy()
    tmpI[np.logical_not(roi)] = 1
    maxI = np.percentile(tmpI, 99, axis=1, keepdims=True)

    I = 255 * I / maxI
    I[I > 255] = 255

    # Smooth out the image with a guassian blur
    I = cv2.GaussianBlur(I, (7, 7), 1)  # This uses matlab symmetric padding

    return I, roi


def locate_perpendicular_lines(I, roi, threshold=2, reduce_points=True):

    # Get the size of the image
    img_size = I.shape

    # Image gradient along either direction
    Gx = cv2.Sobel(I, -1, 1, 0, ksize=3)
    Gy = cv2.Sobel(I, -1, 0, 1, ksize=3)

    G = np.sqrt(Gx * Gx + Gy * Gy)
    mask = G > np.max(G) / threshold

    if reduce_points:
        # We do not need quite so many points, so remove 2/3'rds of them.
        # This is not the same as just increasing the threshold on the mask
        # above
        mask2 = np.zeros(mask.size, dtype=bool)
        mask2[::3] = True
        mask2.shape = mask.T.shape
        mask2 = mask2.T

        mask = mask & mask2 & roi

    # Find the location of the non-zero pixels
    y_idx, x_idx = mask.nonzero()
    y_idx.shape = (y_idx.size, 1)
    x_idx.shape = (x_idx.size, 1)

    # Get the slope of the line at each of these points
    # -- the slope of the line is perpendicular to the gradient
    Th = np.arctan2(Gy[mask], Gx[mask])
    Th[Th < 0] += 2 * pi  # map [0,2*pi]
    Th = Th + pi / 2  # Perpendicular to gradient
    slope = np.tan(Th)  # slope of the lines
    slope.shape = (slope.size, 1)

    # Compute the end points of the lines that conver the image, have the
    # slopes as given by above and intersect the points above
    #
    # There are likely faster/better ways of doing this, but it works

    # Lines starting and ending at x-bounds (but exceeding y bounds)
    x = np.ones((slope.size, 2)) * np.array((1, img_size[1])) - 1
    y = slope * (x - x_idx) + y_idx

    # Lines starting and ending at y-bounds (but exceeding x bounds)
    yr = np.ones((slope.size, 2)) * np.array((1, img_size[0])) - 1
    xr = (yr - y_idx) / slope + x_idx

    # Sort xr/yr so that xr[:,0] is less than xr[:,1] (same order as x)
    sort_ind = xr.argsort(axis=1)
    xr = xr[np.arange(slope.size)[:, None], sort_ind]
    yr = yr[np.arange(slope.size)[:, None], sort_ind]

    # if y is ousize of the y-bounds, then replace row with xr/yr values
    # --start point
    replace = (y[:, 0] < 0) | (y[:, 0] > (img_size[0] - 1))
    x[replace, 0] = xr[replace, 0]
    y[replace, 0] = yr[replace, 0]

    # --end point
    replace = (y[:, 1] < 0) | (y[:, 1] > (img_size[0] - 1))
    x[replace, 1] = xr[replace, 1]
    y[replace, 1] = yr[replace, 1]

    # round all values to nearest integer and cast as integers
    x = np.around(x).astype(np.int_)
    y = np.around(y).astype(np.int_)

    # create arrays with the start point (y,x)_start and the end point
    # (y,x)_end (we do just convert back to the x,y in the accumLines
    # function, but, haveing starting/end points together seems nice.)
    x1 = np.column_stack((y[:, 0], x[:, 0]))
    x2 = np.column_stack((y[:, 1], x[:, 1]))

    return x1, x2


def compute_lines(vote_landscape, roi, previous_VL=None):

    # import matplotlib.pyplot as plt

    img_size = vote_landscape.shape

    VL = cv2.GaussianBlur(vote_landscape, (15, 15), 2)

    current_VL = VL.copy()

    if previous_VL is not None:
        if previous_VL.size != 0:
            VL = np.mean(np.dstack((VL, previous_VL)), axis=2)

    VL[np.logical_not(roi)] = 0

    VLsum0 = VL.sum(axis=0, keepdims=True)
    VLsum0[VLsum0 < 1] = 1

    VL = (VLsum0 > np.percentile(VLsum0, 25)) * VL / VLsum0
    VL = VL > np.percentile(VL, 97)

    y, x = VL.nonzero()
    y = y.astype(np.float32)
    x = x.astype(np.float32)

    x_mid = x.mean()

    y_min = np.around(img_size[0] * 0.6)
    # x0 = np.zeros(2)
    #
    # def fun(x, t, y):
    #     return x[1]*t + x[0] - y

    inrange = x > (x_mid + 10)
    X = np.column_stack((np.ones(np.count_nonzero(inrange)), x[inrange] + 1))

    # This returns the intercept and slope of the line
    wr = lstsq(X, y[inrange] + 1)[0]
    # wr = least_squares(fun, x0, args=(x[inrange], y[inrange]))
    # print(wr)

    # ends points of right line
    yr = np.array([img_size[0] - 1, y_min])
    xr = np.around((yr - wr[0]) / wr[1])

    inrange = x < (x_mid - 10)
    X = np.column_stack((np.ones(np.count_nonzero(inrange)), x[inrange] + 1))
    # This returns the intercept and slope of the line
    wl = lstsq(X, y[inrange] + 1)[0]
    # wl = least_squares(fun, x0, args=(x[inrange], y[inrange]))[0]
    # ends points of right line
    yl = np.array([img_size[0] - 1, y_min])
    xl = np.around((yl - wl[0]) / wl[1])

    ll = np.column_stack((yl, xl))  # left lane [p1; p2]
    rl = np.column_stack((yr, xr))  # right lane [p1; p2]

    return ll, rl, current_VL, VL


def create_ROI(img_size, vertices):
    # Vertices should be an array of x,y values that give the fraction
    # the image.
    roi = np.zeros(img_size)
    vrts = np.around(
        vertices * (img_size[1] - 1, img_size[0] - 1)).astype(np.int32)
    cv2.fillPoly(roi, vrts, 1)
    roi = roi > 0
    return roi


def find_frames_lane_lines(I, roi, history):

    # Get image size
    orig_img_size = I.shape[0:2]

    # Define some constants
    # The imageas do need to be so large, so scale them down to this size if
    # larger
    WORKING_IMAGE_WIDTH = 400
    RESIZE_FACTOR = np.around(WORKING_IMAGE_WIDTH / orig_img_size[1], 2)
    # RESIZED_IMAGE = (int(np.around(orig_img_size[1]*RESIZE_FACTOR)), int(400))
    # Resize image, adjust contrast in ROI along rows, and smooth out the
    # image with a gaussian blur (sig=2)
    I, roi = pre_process_image(I, roi, RESIZE_FACTOR)

    # Get image size
    img_size = I.shape[0:2]

    # Find the end points for the lines perpendicular to large
    # gradients of I in ROI
    x1, x2 = locate_perpendicular_lines(I, roi)

    # For each of the lines computed above, accumulate the pixels they
    # cross over in an image, this will make all straight lines much easier
    # to threshold and the dotted lines will be automatically connected.
    A = accumLines(x1, x2, img_size)

    # Compute the end points of the lines and return the processed
    # accumulated matrix for saving
    ll, rl, VL, mask = compute_lines(A, roi, history.VL)

    # Rescale the computed lines to the original image size
    ll = ll / RESIZE_FACTOR
    rl = rl / RESIZE_FACTOR

    # Add new observation to history
    history.add(VL, ll, rl)
    # saving the mask only because it will be helpful to plot while describing
    # the method
    history.mask = mask

File: findLaneLines.m
function [results, count] = findLaneLines(I,history,roi)


RESIZE_FACTOR = round(400/size(I,2),2);%0.4;

I = mean(I(:,:,1:2),3); % This converts I to double

I = imresize(I,RESIZE_FACTOR,'bilinear','Antialiasing',false);
roi = imresize(roi,RESIZE_FACTOR,'bilinear','Antialiasing',false)>0;

y_min = find(sum(roi,2),1,'first');


tmpI = I;
tmpI(~roi) = Inf;
minI = prctile(tmpI',30)';
% minI(any(roi,2)) = minI_mask(any(roi,2));
I = I - minI;
I(I<0) = 0;

tmpI = I;
tmpI(~roi) = 1;
maxI = prctile(tmpI',99)';

I = 255*(I./maxI);%prctile(I(roi),95)));
I(I>255) = 255;

I = imfilter(I,fspecial('gaussian',7,1),'replicate');



h = -fspecial('sobel'); % Align mask correctly along the x- and y- axes
Gx = imfilter(I,h','replicate');
if nargout > 1
    Gy = imfilter(I,h,'replicate');
end

% [Gx,Gy] = gradient(I);
G = sqrt(Gx.^2 + Gy.^2);


% Create the set of pixels that we will create lines on
mask = G > max(G(:))/2;

% mask = edge(I,'canny',[100,180]/255);
mask = logical(mask .* roi);

mask_mask = false(numel(mask),1);
mask_mask(1:3:end) = true;
mask_mask = reshape(mask_mask,size(mask));
mask = mask & mask_mask;



[i_y,i_x] = find(mask);

% At each pixel, compute the slope of the line.
Th = atan2(Gy(mask),Gx(mask)); % Angle of gradient
Th(Th<0) = Th(Th<0) + 2*pi;
Th = Th + pi/2; % The line is perpendicular to the gradient
m = tan(Th); % copmute the slope

% The the end points of each line bounded by the image.
x = [1,size(I,2)] .* ones(numel(i_x),1);
y = m.*x - m.*i_x + i_y;

y2 = [1, size(I,1)] .* ones(numel(i_x),1);
x2 = y2./m - i_y./m + i_x;

[x2,si] = sort(x2,2);
y2 = y2((si-1)*numel(i_x) + (1:numel(i_x))');

replace = y(:,1) < 1 | y(:,1) > size(I,1);
x(replace,1) = x2(replace,1);
y(replace,1) = y2(replace,1);

replace = y(:,2) > size(I,1) | y(:,2) < 1;
x(replace,2) = x2(replace,2);
y(replace,2) = y2(replace,2);

% Round end points to nearest integer and store as end points.
x = round(x);
y = round(y);

x1 = [y(:,1), x(:,1)];
x2 = [y(:,2), x(:,2)];

% Accumulate the lines into an image
count = accumLines(x1,x2,size(I),4);


% Smooth out the counts
count = imfilter(count,fspecial('gaussian',14,2));


% Normalize the counts using the mask
% mask2 = imdilate(mask,ones(3));
[~,vanish_idx] = max(count(:));%./(1+mask2(:)));
[y_max,x_max] = ind2sub(size(count),vanish_idx);


results = [];
results.count = count;

if ~isempty(history.count)
    count = mean(cat(3,count,history.count),3);
end

count = count .* roi;

count_mean = sum(count,1);
count = (count_mean > prctile(count_mean,25)) .* count ./ (count_mean);


% Threshold the image
count = count > prctile(count(:),97); %This should require approximately 10 lines to pass through each point.
% count = imclose(count,strel('disk',2)); % close the image to connect the two sides of the lane lines


% Grap the
% [H,T,R] = hough(bwmorph(count,'thin',Inf),'RhoResolution',1,'Theta',-90:2:89);
% P  = houghpeaks(H,5,'threshold',ceil(0.2*max(H(:))));
% lines = houghlines(count,T,R,P,'FillGap',5,'MinLength',round(max(size(I))/3));
% 
% % figure(2)
% % imshow(count + bwmorph(count,'thin',5),[0,2])
% 
% th = [lines.theta];
% thp = th + 360 * (th<0);
% [~,idx1] = min(thp); % left lane
% thn = th - 360 * (th>0);
% [~,idx2] = max(thn); % right lane
% 
% 
% right = [lines(idx2).point1; lines(idx2).point2];
% left = [lines(idx1).point1; lines(idx1).point2];
% 
% 
% 
% if ~isempty(right)
%     [~,right_ord] = sort(right(:,2),1,'descend');
%     right = right(right_ord,:);% + [left_bnd,top_bnd];
% end
% 
% if ~isempty(left)
%     [~,left_ord] = sort(left(:,2),1,'descend');
%     left = left(left_ord,:);% + [left_bnd,top_bnd];
% end



[y,x] = find(count);

x_mid = mean(x);

inrange = x > x_mid+10;

% inrange = (x < max(right(:,1))) & (x > min(right(:,1)));
right_param = [ones(sum(inrange),1), x(inrange)]\y(inrange);
right(:,2) = [size(I,1); y_min];
right(:,1) = round((right(:,2) - right_param(1))/right_param(2));


inrange = x < x_mid-10;
% inrange = (x < max(left(:,1))) & (x > min(left(:,1)));
left_param = [ones(sum(inrange),1), x(inrange)]\y(inrange);
left(:,2) = [size(I,1); y_min];
left(:,1) = round((left(:,2) - left_param(1))/left_param(2));


top = [x_max,y_max];

right = right / RESIZE_FACTOR;
left = left / RESIZE_FACTOR;
top = top / RESIZE_FACTOR;


results.left = left;
results.right = right;
results.vanish = top;

end
